### DISCUSSION TITLE
How to update the webview when data changed ?

### DISCUSSION THREAD

**Discussion by liukaimingzzx** [Kitware/trame#237] [Q&A] [answered]
I want to implement a web component to display log informations. The codes are as followings:
```python
with vuetify.VSnackbar(
        v_model="show_log",
        timeout=-1,
        color="white",
        min_width=600,
        max_width=800,
        classes="ma-auto pa-0",

):
    with vuetify.VCard(color="white", classes="mx-auto px-0") as nn:
        vuetify.VCardTitle("Log Information", style="color: black; background-color: grey", )
        with vuetify.VCardText(
                style="white-space: pre-line; max-height: 200px; overflow: auto; color: black",
                v_model=("log_info", log_info)) as vc:
            str=''
            for i in log_info:
                str += i+'\n'
            vc.set_text(str)
```

The  variable ```log_info```  is a python list which contains log information like ```"[2021-07-27 10:00:00] info:xxxxxxxx", "[2021-07-27 11:00:00] info:xxxxxxxx"```. With the running of the program, more log information will be added to```log_info```

I would like to implement a feature where the log component on the web page will be updated when the logging information in the Python list is updated. Could you please give me some suggestions?

**Comment by jourdain** [comment]
With trame the UI definition is aimed to be executed once and let the dynamic aspect of its template play out when state variables are getting updated. In your case, you don't want to resend the full log every time a new line is added (or even the full HTML). Therefore a better approach would be to use something line [trame-xterm](https://github.com/Kitware/trame-xterm) where you can send one line at a time. You even have an example [doing it here](https://github.com/Kitware/trame-xterm/blob/master/examples/readonly/app.py). Just don't forget to add it to your python dependency of your project. 

HTH

**Reply by liukaimingzzx** [reply]
Thanks for your suggestion, we will try trame-xterm later. 
But we still don't know why the codes below could not update the 'Log Information' card.
1. Fisrt, we put the list ```log_info``` into state
```
state.log_info = ["[2021-07-27 10:00:00] info:xxxxxxxx", "[2021-07-27 11:00:00] info:xxxxxxxx",
                  "[2021-07-27 12:00:00] info:xxxxxxxx"]
``` 
2. Then we wirte the UI element of the 'Log Information' Card, We use ```v-for``` to iterate each item in ```log_info``` and dynamically render them.:
```
with vuetify.VCard(color="white", classes="mx-auto px-0"): 
    vuetify.VCardTitle("Log Information", style="color: black; background-color: grey")
    with vuetify.VCardText(
            style="white-space: pre-line; max-height: 200px; overflow: auto; color: black",
    ):
        with html.Div(v_for="item in log_info", key="item"):
            html.Span("{{item}}")
```
3. The we wirte a handler, when the variable ```show-log```changes, we append a new item to ```log_info```:
```
@state.change("show_log")
def s_change(log_info, **kwargs):
    log_info.append("new info")
```

These codes could modify the ```log_info```, but could not able to update the log information in the web ui card.

Could you help me identify the issue with this piece of code and correct it?  Thanks.

**Reply by jourdain** [reply, marked as answer]
Because you are keeping the same list on the server, the server does not detect the change automatically but you can help it by doing the following

```python
@state.change("show_log")
def s_change(log_info, **kwargs):
    log_info.append("new info")
    state.dirty('log_info') # https://trame.readthedocs.io/en/latest/core.state.html#trame_server.state.State.dirty
```

Also for your template it will be better to do it that way

```python
with html.Div(v_for="item, idx in log_info", key="idx"):
     html.Span("{{item}}")
```

**Reply by liukaimingzzx** [reply]
Thank you sir, your suggestions help a lot !
